File: elixir/change/change.exs
defmodule Change do
  @doc """
    Determine the least number of coins to be given to the user such
    that the sum of the coins' value would equal the correct amount of change.
    It returns :error if it is not possible to compute the right amount of coins.
    Otherwise returns the tuple {:ok, map_of_coins}

    ## Examples

      iex> Change.generate(3, [5, 10, 15])
      :error

      iex> Change.generate(18, [1, 5, 10])
      {:ok, %{1 => 3, 5 => 1, 10 => 1}}

  """
  @spec generate(integer, list) :: {:ok, map} | :error
  def generate(amount, []), do: :error
  def generate(amount, [head|_tail]) when amount < head, do: :error
  def generate(amount, values) do
    map_of_coins = values |> Enum.sort |> Enum.reverse |> do_generate(amount, %{})
    {:ok, map_of_coins}
  end

  defp do_generate([], _amount, acc), do: acc
  defp do_generate([head|tail], amount, acc) do
    { coins, remainder } = divrem(amount, head)
    do_generate(tail, remainder, Map.put(acc, head, coins))
  end

  defp divrem(left, right), do: { div(left,right), rem(left,right) }
end

File: go/change/change.go
package change

import "errors"

func Change(coins []int, target int) ([]int, error) {
	var results []int

	for i := len(coins); i > 0; i-- {
		change, err := makeChange(coins[:i], target)
		if err != nil {
			continue
		}

		if results == nil || len(change) < len(results) {
			results = change
		}
	}

	if results == nil {
		return nil, errors.New("Exact change can't be given")
	}

	return results, nil
}

func makeChange(coins []int, target int) ([]int, error) {
	change := make([]int, 0, len(coins))

	for i := len(coins) - 1; i >= 0; i-- {
		for target >= coins[i] {
			if target-coins[i] != 0 && target-coins[i] < coins[0] {
				break
			}

			change = append([]int{coins[i]}, change...)
			target -= coins[i]

			if i > 0 && target%coins[i] != 0 && target%coins[i-1] == 0 {
				break
			}
		}
	}

	if target != 0 {
		return nil, errors.New("Exact change can't be given")
	}

	return change, nil
}
